core/server/src/test/java/alluxio/master/file/ConcurrentFileSystemMasterTest.java

```diff
@@ -0,0 +1,548 @@
+ * path as the under storage system.
+ *
+ * The tests validate the correctness of concurrent operations, ie. no corrupted/partial state is
+ * exposed, through a series of concurrent operations followed by verification of the final
+ * state, or inspection of the in-progress state as the operations are carried out.
+ *
+ * The tests also validate that operations are concurrent by injecting a short sleep in the
+ * critical code path. Tests will timeout if the critical section is performed serially.
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(UnderFileSystem.Factory.class)
+public class ConcurrentFileSystemMasterTest {
+  private static final String TEST_USER = "test";
+  private static final long SLEEP_MS = Constants.SECOND_MS;
+  /**
+   * Options to mark a created file as persisted. Note that this does not actually persist the
+   * file but flag the file to be treated as persisted, which will invoke ufs operations.
+   */
+  private static CreateFileOptions sCreatePersistedFileOptions;
+
+  private BlockMaster mBlockMaster;
+  private FileSystemMaster mFileSystemMaster;
+
+  @Rule
+  public AuthenticatedUserRule mAuthenticatedUser = new AuthenticatedUserRule(TEST_USER);
+
+  /**
+   * Sets up the dependencies before a test runs.
+   */
+  @Before
+  public void before() throws Exception {
+    LoginUserTestUtils.resetLoginUser();
+    GroupMappingServiceTestUtils.resetCache();
+    Configuration.set(PropertyKey.SECURITY_LOGIN_USERNAME, TEST_USER);
+    // Set umask "000" to make default directory permission 0777 and default file permission 0666.
+    Configuration.set(PropertyKey.SECURITY_AUTHORIZATION_PERMISSION_UMASK, "000");
+    sCreatePersistedFileOptions = CreateFileOptions.defaults().setPersisted(true);
+
+    // Cannot use multiple rules with Powermock, so use Java tmp dir
+    String ufsAddress = Files.createTempDir().getAbsolutePath();
+    String journalAddress = Files.createTempDir().getAbsolutePath();
+
+    Configuration.set(PropertyKey.UNDERFS_ADDRESS, ufsAddress);
+
+    JournalFactory journalFactory = new JournalFactory.ReadWrite(journalAddress);
+    mBlockMaster = new BlockMaster(journalFactory);
+    ExecutorService service =
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("FileSystemMasterTest-%d", true));
+    mFileSystemMaster = new FileSystemMaster(mBlockMaster, journalFactory,
+        ExecutorServiceFactories.constantExecutorServiceFactory(service));
+
+    mBlockMaster.start(true);
+    mFileSystemMaster.start(true);
+
+    PowerMockito.mockStatic(UnderFileSystem.Factory.class);
+    PowerMockito.when(UnderFileSystem.Factory.get(Mockito.anyString())).thenAnswer(
+        new Answer<UnderFileSystem>() {
+          @Override
+          public UnderFileSystem answer(InvocationOnMock invocationOnMock) throws Throwable {
+            AlluxioURI uri = new AlluxioURI((String) invocationOnMock.getArguments()[0]);
+            return new SleepingLocalUnderFileSystem(uri, SLEEP_MS);
```

Comment:
As of now, this ufs is fairly specialized for these tests, a fully "operational" sleeping file system would require some more consideration of which commands are no-oped and which ones need to execute (ie. `resolveUri`)